File: packages/sdk/pact-go/invariants/signing.go
package invariants

import (
	"crypto/ed25519"
	"encoding/hex"
	"strings"
)

type SignedMessage struct {
	CanonicalJSON string
	PublicKeyHex  string
	SignatureHex  string
}

func IsValidEd25519PublicKeyHex(value string) bool {
	_, err := decodeFixedHex(value, 32, "invalid_public_key")
	return err == nil
}

func IsValidEd25519SignatureHex(value string) bool {
	_, err := decodeFixedHex(value, 64, "invalid_signature")
	return err == nil
}

func PublicKeyHexMatches(left string, right string) bool {
	return normalizeHex(left) == normalizeHex(right)
}

func SignUTF8MessageEd25519(input string, seedHex string) (SignedMessage, error) {
	seed, err := decodeFixedHex(seedHex, 32, "invalid_hex")
	if err != nil {
		return SignedMessage{}, err
	}
	privateKey := ed25519.NewKeyFromSeed(seed)
	publicKey := privateKey.Public().(ed25519.PublicKey)
	signature := ed25519.Sign(privateKey, []byte(input))
	return SignedMessage{
		PublicKeyHex: hex.EncodeToString(publicKey),
		SignatureHex: hex.EncodeToString(signature),
	}, nil
}

func VerifyUTF8MessageEd25519(input string, publicKeyHex string, signatureHex string) (bool, error) {
	publicKey, err := decodeFixedHex(publicKeyHex, 32, "invalid_public_key")
	if err != nil {
		return false, err
	}
	signature, err := decodeFixedHex(signatureHex, 64, "invalid_signature")
	if err != nil {
		return false, err
	}
	return ed25519.Verify(ed25519.PublicKey(publicKey), []byte(input), signature), nil
}

func SignJSONStringEd25519(input string, seedHex string) (SignedMessage, error) {
	canonicalJSON, err := CanonicalizeJSONString(input)
	if err != nil {
		return SignedMessage{}, err
	}
	signed, err := SignUTF8MessageEd25519(canonicalJSON, seedHex)
	if err != nil {
		return SignedMessage{}, err
	}
	signed.CanonicalJSON = canonicalJSON
	return signed, nil
}

func VerifyJSONStringSignatureEd25519(input string, publicKeyHex string, signatureHex string) (bool, error) {
	canonicalJSON, err := CanonicalizeJSONString(input)
	if err != nil {
		return false, err
	}
	return VerifyUTF8MessageEd25519(canonicalJSON, publicKeyHex, signatureHex)
}

func normalizeHex(value string) string {
	normalized := strings.TrimPrefix(strings.TrimPrefix(value, "0x"), "0X")
	return strings.ToLower(normalized)
}

func decodeFixedHex(value string, expectedBytes int, code string) ([]byte, error) {
	normalized := normalizeHex(value)
	if len(normalized) != expectedBytes*2 {
		return nil, newInvariantError(code, "value has an unexpected hexadecimal length")
	}
	decoded, err := hex.DecodeString(normalized)
	if err != nil {
		return nil, newInvariantError(code, "value is not valid hexadecimal")
	}
	return decoded, nil
}

File: packages/sdk/chio-ts/src/invariants/signing.ts
import {
  isValidEd25519PublicKeyHex,
  isValidEd25519SignatureHex,
  publicKeyHexMatches,
  signEd25519Message,
  verifyEd25519Signature,
} from "./crypto.ts";
import { canonicalizeJsonString } from "./json.ts";

export interface Utf8MessageSignature {
  public_key_hex: string;
  signature_hex: string;
}

export interface CanonicalJsonSignature extends Utf8MessageSignature {
  canonical_json: string;
}

export function isValidPublicKeyHex(value: string): boolean {
  return isValidEd25519PublicKeyHex(value);
}

export function isValidSignatureHex(value: string): boolean {
  return isValidEd25519SignatureHex(value);
}

export function signUtf8MessageEd25519(input: string, seedHex: string): Utf8MessageSignature {
  return signEd25519Message(input, seedHex);
}

export function verifyUtf8MessageEd25519(
  input: string,
  publicKeyHex: string,
  signatureHex: string,
): boolean {
  return verifyEd25519Signature(input, publicKeyHex, signatureHex);
}

export function signJsonStringEd25519(input: string, seedHex: string): CanonicalJsonSignature {
  const canonical_json = canonicalizeJsonString(input);
  const signed = signEd25519Message(canonical_json, seedHex);
  return {
    canonical_json,
    public_key_hex: signed.public_key_hex,
    signature_hex: signed.signature_hex,
  };
}

export function verifyJsonStringSignatureEd25519(
  input: string,
  publicKeyHex: string,
  signatureHex: string,
): boolean {
  return verifyEd25519Signature(canonicalizeJsonString(input), publicKeyHex, signatureHex);
}

export { publicKeyHexMatches };
